File: scripts/sonar_occupancy_mapper.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import rospy
import numpy as np
import tf2_ros
import tf2_py as tf2
from sensor_msgs.msg import PointCloud2
from sensor_msgs import point_cloud2 as pc2
from nav_msgs.msg import OccupancyGrid, Odometry
import math

class SonarOccupancyMapper:
    def __init__(self):
        rospy.init_node('sonar_occupancy_mapper', anonymous=True)
        
        # --- 参数配置 ---
        self.map_frame = "world"
        self.robot_base_frame = "rexrov2/base_link"
        
        self.map_res = 0.2
        self.map_width_m = 300.0
        self.map_height_m = 300.0
        self.map_origin_x = -150.0
        self.map_origin_y = -150.0
        
        # [修改] 问题3: 移除硬编码 Z 轴，设置为 0.0 或参数化
        self.map_z_level = 0.0 
        
        self.sonar_max_range = 16.0 
        self.sonar_fov_rad = np.deg2rad(95.0) 
        
        # 贝叶斯/计数更新参数
        self.hit_increment = 30.0
        self.max_occupancy = 100.0
        
        # 衰减参数 (仅在可视区域内衰减，速度快一点)
        self.base_decay_rate = 5.0  
        self.velocity_decay_factor = 1.0 
        
        # 插值/膨胀参数
        self.dilation_radius = 1
        
        # 姿态过滤阈值 (度)
        self.max_tilt_deg = 5.0
        
        # --- 内部变量 ---
        self.grid_shape = (int(self.map_height_m / self.map_res), 
                           int(self.map_width_m / self.map_res))
        
        self.occupancy_grid = np.zeros(self.grid_shape, dtype=np.float32)
        self.current_speed = 0.0
        
        # [修改] 问题2: 初始化为 None，防止第一帧 dt 过大
        self.last_update_time = None
        
        # [修改] 问题6: 性能优化 - 预计算查找表 (Cache)
        self._init_precomputed_tables()

        # --- TF 监听器 ---
        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)
        
        # --- 订阅与发布 ---
        rospy.Subscriber('/sonar_cloud_body', PointCloud2, self.cloud_callback, queue_size=1, buff_size=2**24)
        rospy.Subscriber('/rexrov2/pose_gt', Odometry, self.odom_callback, queue_size=1)
        
        self.map_pub = rospy.Publisher('/projected_sonar_map', OccupancyGrid, queue_size=1)
        
        rospy.loginfo(f"Sonar Mapper Started. Map Size: {self.grid_shape}")

    def _init_precomputed_tables(self):
        """
        [修改] 问题6: 预计算相对坐标和角度，避免在回调中进行昂贵的 meshgrid 和 arctan2 运算
        """
        # 计算 ROI 半径对应的网格数
        self.range_indices = int(self.sonar_max_range / self.map_res) + 1
        size = 2 * self.range_indices + 1
        
        # 生成相对中心的网格 (像素坐标)
        x_range = np.arange(-self.range_indices, self.range_indices + 1)
        y_range = np.arange(-self.range_indices, self.range_indices + 1)
        self.cache_grid_x, self.cache_grid_y = np.meshgrid(x_range, y_range)
        
        # 转换为相对物理坐标
        rel_world_x = self.cache_grid_x * self.map_res
        rel_world_y = self.cache_grid_y * self.map_res
        
        # 预计算距离平方 (用于距离判断)
        self.cache_dist_sq = rel_world_x**2 + rel_world_y**2
        
        # 预计算相对角度 (用于 FOV 判断)
        self.cache_angles = np.arctan2(rel_world_y, rel_world_x)
        
        rospy.loginfo("Performance cache initialized.")

    def odom_callback(self, msg):
        vx = msg.twist.twist.linear.x
        vy = msg.twist.twist.linear.y
        self.current_speed = np.sqrt(vx**2 + vy**2)

    def cloud_callback(self, cloud_msg):
        current_time = rospy.Time.now()
        
        # [修改] 问题2: 处理第一帧逻辑
        if self.last_update_time is None:
            self.last_update_time = current_time
            # 第一帧只做初始化，不计算 dt，避免巨大的 decay
            return

        dt = (current_time - self.last_update_time).to_sec()
        if dt < 0: dt = 0.0
        self.last_update_time = current_time

        # =========================================================
        # 步骤 1: 严格姿态检查
        # =========================================================
        try:
            # 显式查询机器人基座(Base Link)的变换，而非声纳的变换
            # [修改] 问题4: 增加超时时间到 0.5s，提高系统高负载下的鲁棒性
            base_trans = self.tf_buffer.lookup_transform(
                self.map_frame, 
                self.robot_base_frame, 
                cloud_msg.header.stamp, 
                rospy.Duration(0.5) 
            )
            
            bq_x = base_trans.transform.rotation.x
            bq_y = base_trans.transform.rotation.y
            bq_z = base_trans.transform.rotation.z
            bq_w = base_trans.transform.rotation.w
            
            # 计算基座的 Roll 和 Pitch
            base_roll, base_pitch, _ = self.get_euler_from_quaternion(bq_x, bq_y, bq_z, bq_w)
            # 检查是否超过阈值
            max_tilt_rad = np.deg2rad(self.max_tilt_deg)
            if abs(base_roll) > max_tilt_rad or abs(base_pitch) > max_tilt_rad:
                # 姿态不稳，跳过更新，防止地图模糊
                # 此时仅发布旧地图以保持 Topic 活跃
                self.publish_map(cloud_msg.header.stamp)
                return

        except (tf2.LookupException, tf2.ConnectivityException, tf2.ExtrapolationException) as e:
            rospy.logwarn_throttle(2.0, f"TF Lookup Failed (Base): {e}")
            return

        # =========================================================
        # 步骤 2: 获取声纳传感器位姿
        # =========================================================
        try:
            # [修改] 问题4: 增加超时时间
            sensor_trans = self.tf_buffer.lookup_transform(
                self.map_frame, 
                cloud_msg.header.frame_id, 
                cloud_msg.header.stamp, 
                rospy.Duration(0.5)
            )
        except (tf2.LookupException, tf2.ConnectivityException, tf2.ExtrapolationException) as e:
            rospy.logwarn_throttle(2.0, f"TF Lookup Failed (Sensor): {e}")
            return

        # 提取传感器参数
        t_x = sensor_trans.transform.translation.x
        t_y = sensor_trans.transform.translation.y
        t_z = sensor_trans.transform.translation.z
        
        q_x = sensor_trans.transform.rotation.x
        q_y = sensor_trans.transform.rotation.y
        q_z = sensor_trans.transform.rotation.z
        q_w = sensor_trans.transform.rotation.w
        
        _, _, sensor_yaw = self.get_euler_from_quaternion(q_x, q_y, q_z, q_w)

        # =========================================================
        # 步骤 3: 处理点云 (Hit Update)
        # =========================================================
        gen = pc2.read_points(cloud_msg, field_names=("x", "y", "z"), skip_nans=True)
        points_sensor = np.array(list(gen), dtype=np.float32)
        
        rotation_matrix = self.quaternion_to_matrix(q_x, q_y, q_z, q_w)
        translation = np.array([t_x, t_y, t_z])

        if points_sensor.size > 0:
            # 坐标转换: Sensor -> World
            points_world = np.dot(points_sensor, rotation_matrix.T) + translation
            
            # 映射到网格索引
            j_indices = ((points_world[:, 0] - self.map_origin_x) / self.map_res).astype(int)
            i_indices = ((points_world[:, 1] - self.map_origin_y) / self.map_res).astype(int)
            
            # 过滤有效范围
            valid_mask = (i_indices >= 0) & (i_indices < self.grid_shape[0]) & \
                         (j_indices >= 0) & (j_indices < self.grid_shape[1])
            
            valid_i = i_indices[valid_mask]
            valid_j = j_indices[valid_mask]

            if len(valid_i) > 0:
                current_hits = np.zeros(self.grid_shape, dtype=bool)
                current_hits[valid_i, valid_j] = True
                dilated_hits = self.dilate_mask(current_hits, self.dilation_radius)
                
                self.occupancy_grid[dilated_hits] += self.hit_increment
                self.occupancy_grid[dilated_hits] = np.clip(self.occupancy_grid[dilated_hits], 0, self.max_occupancy)

        # =========================================================
        # 步骤 4: 衰减更新 (使用优化后的逻辑，保留原始衰减算法)
        # =========================================================
        self.apply_fov_decay_optimized(dt, t_x, t_y, sensor_yaw)

        self.publish_map(cloud_msg.header.stamp)

    def apply_fov_decay_optimized(self, dt, sensor_x, sensor_y, sensor_yaw):
        """
        [修改] 问题6: 使用预计算的 Cache 进行快速衰减计算
        逻辑保持不变：在 FOV 范围内进行衰减。
        """
        if dt <= 0: return

        # 1. 计算传感器在 Grid 中的整数索引
        cx = int((sensor_x - self.map_origin_x) / self.map_res)
        cy = int((sensor_y - self.map_origin_y) / self.map_res)

        # 2. 确定裁剪区域 (处理地图边界)
        # 这里的 min/max 是相对于 Cache 数组的索引
        pad = self.range_indices
        
        # 计算在地图上的绝对范围
        min_map_x = cx - pad
        max_map_x = cx + pad + 1
        min_map_y = cy - pad
        max_map_y = cy + pad + 1
        
        # 计算在 Cache 上的切片范围
        cache_min_x, cache_max_x = 0, self.cache_grid_x.shape[1]
        cache_min_y, cache_max_y = 0, self.cache_grid_y.shape[0]

        # 边界裁剪逻辑
        if min_map_x < 0:
            cache_min_x = -min_map_x
            min_map_x = 0
        if min_map_y < 0:
            cache_min_y = -min_map_y
            min_map_y = 0
        if max_map_x > self.grid_shape[1]:
            diff = max_map_x - self.grid_shape[1]
            cache_max_x -= diff
            max_map_x = self.grid_shape[1]
        if max_map_y > self.grid_shape[0]:
            diff = max_map_y - self.grid_shape[0]
            cache_max_y -= diff
            max_map_y = self.grid_shape[0]

        # 如果完全在地图外，直接返回
        if min_map_x >= max_map_x or min_map_y >= max_map_y:
            return

        # 3. 提取 Cache 切片 (避免重复计算 distance 和 angle)
        roi_dist_sq = self.cache_dist_sq[cache_min_y:cache_max_y, cache_min_x:cache_max_x]
        roi_angles = self.cache_angles[cache_min_y:cache_max_y, cache_min_x:cache_max_x]

        # 4. 计算角度差 (只需要做减法)
        # 将传感器 Yaw 减去，得到相对于传感器朝向的角度
        delta_angle = roi_angles - sensor_yaw
        delta_angle = (delta_angle + np.pi) % (2 * np.pi) - np.pi

        # 5. 生成 Mask
        decay_max_range = self.sonar_max_range + 1.0
        decay_fov_rad = self.sonar_fov_rad + np.deg2rad(5.0)
        
        fov_mask = (roi_dist_sq < decay_max_range**2) & \
                   (np.abs(delta_angle) < decay_fov_rad / 2.0)

        # 6. 应用衰减 (直接操作切片)
        decay_amount = (self.base_decay_rate + self.current_speed * self.velocity_decay_factor) * dt
        
        # 获取地图上的 ROI 视图
        roi_grid = self.occupancy_grid[min_map_y:max_map_y, min_map_x:max_map_x]
        
        # 应用更新
        roi_grid[fov_mask] -= decay_amount
        roi_grid[fov_mask] = np.clip(roi_grid[fov_mask], 0, self.max_occupancy)

    # --- 辅助函数保持不变 ---
    def dilate_mask(self, mask, radius):
        if radius <= 0: return mask
        dilated = mask.copy()
        rows, cols = mask.shape
        for dy in range(-radius, radius + 1):
            for dx in range(-radius, radius + 1):
                if dy == 0 and dx == 0: continue
                s_y0, s_y1 = max(0, -dy), min(rows, rows - dy)
                s_x0, s_x1 = max(0, -dx), min(cols, cols - dx)
                d_y0, d_y1 = max(0, dy), min(rows, rows + dy)
                d_x0, d_x1 = max(0, dx), min(cols, cols + dx)
                dilated[d_y0:d_y1, d_x0:d_x1] |= mask[s_y0:s_y1, s_x0:s_x1]
        return dilated

    def quaternion_to_matrix(self, x, y, z, w):
        R = np.array([
            [1 - 2*(y**2 + z**2), 2*(x*y - z*w),     2*(x*z + y*w)],
            [2*(x*y + z*w),       1 - 2*(x**2 + z**2), 2*(y*z - x*w)],
            [2*(x*z - y*w),       2*(y*z + x*w),       1 - 2*(x**2 + y**2)]
        ])
        return R

    def get_euler_from_quaternion(self, x, y, z, w):
        sinr_cosp = 2 * (w * x + y * z)
        cosr_cosp = 1 - 2 * (x * x + y * y)
        roll = math.atan2(sinr_cosp, cosr_cosp)
        sinp = 2 * (w * y - z * x)
        if abs(sinp) >= 1:
            pitch = math.copysign(math.pi / 2, sinp)
        else:
            pitch = math.asin(sinp)
        siny_cosp = 2 * (w * z + x * y)
        cosy_cosp = 1 - 2 * (y * y + z * z)
        yaw = math.atan2(siny_cosp, cosy_cosp)
        return roll, pitch, yaw

    def publish_map(self, stamp):
        grid_msg = OccupancyGrid()
        grid_msg.header.stamp = stamp
        grid_msg.header.frame_id = self.map_frame
        grid_msg.info.resolution = self.map_res
        grid_msg.info.width = self.grid_shape[1]
        grid_msg.info.height = self.grid_shape[0]
        grid_msg.info.origin.position.x = self.map_origin_x
        grid_msg.info.origin.position.y = self.map_origin_y
        # [修改] 问题3: 使用配置好的 Z 高度
        grid_msg.info.origin.position.z = self.map_z_level
        grid_msg.info.origin.orientation.w = 1.0
        
        # [修改] 问题5: 显式 round 和 clip，确保类型转换安全
        # 1. Clip 限制在 0-100
        # 2. Round 四舍五入，而不是默认的 floor
        # 3. 转换为 int8
        final_data = np.clip(self.occupancy_grid, 0, 100)
        final_data = np.round(final_data).astype(np.int8)
        
        grid_msg.data = final_data.flatten().tobytes()
        self.map_pub.publish(grid_msg)

if __name__ == '__main__':
    try:
        node = SonarOccupancyMapper()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass

File: src/sonar_occupancy_mapper.cpp
#include <ros/ros.h>
#include <tf2_ros/transform_listener.h>
#include <tf2_eigen/tf2_eigen.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#include <sensor_msgs/PointCloud2.h>
#include <nav_msgs/OccupancyGrid.h>
#include <nav_msgs/Odometry.h>
#include <pcl_conversions/pcl_conversions.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>
#include <pcl_ros/transforms.h>
#include <vector>
#include <cmath>
#include <algorithm>

class SonarOccupancyMapper {
public:
    SonarOccupancyMapper() : tf_listener_(tf_buffer_) {
        ros::NodeHandle nh;
        ros::NodeHandle pnh("~");

        // --- 参数配置 ---
        pnh.param<std::string>("map_frame", map_frame_, "world");
        pnh.param<std::string>("robot_base_frame", robot_base_frame_, "rexrov2/base_link");
        
        pnh.param("map_res", map_res_, 0.2f);
        pnh.param("map_width_m", map_width_m_, 300.0f);
        pnh.param("map_height_m", map_height_m_, 300.0f);
        pnh.param("map_origin_x", map_origin_x_, -150.0f);
        pnh.param("map_origin_y", map_origin_y_, -150.0f);

        pnh.param("sonar_max_range", sonar_max_range_, 16.0f);
        float fov_deg;
        pnh.param("sonar_fov_deg", fov_deg, 95.0f);
        sonar_fov_rad_ = fov_deg * M_PI / 180.0f;

        pnh.param("hit_increment", hit_increment_, 30.0f);
        pnh.param("max_occupancy", max_occupancy_, 100.0f);
        
        pnh.param("base_decay_rate", base_decay_rate_, 3.0f);
        pnh.param("velocity_decay_factor", velocity_decay_factor_, 1.0f);
        
        pnh.param("dilation_radius", dilation_radius_, 1);
        pnh.param("max_tilt_deg", max_tilt_deg_, 5.0f);

        // --- 内部变量初始化 ---
        grid_width_ = static_cast<int>(map_width_m_ / map_res_);
        grid_height_ = static_cast<int>(map_height_m_ / map_res_);
        grid_size_ = grid_width_ * grid_height_;
        
        // 使用 float 存储以保持精度，发布时转 int8
        occupancy_grid_.resize(grid_size_, 0.0f); 

        current_speed_ = 0.0f;
        last_update_time_ = ros::Time::now();

        // --- 订阅与发布 ---
        sub_cloud_ = nh.subscribe("/sonar_cloud_body", 1, &SonarOccupancyMapper::cloudCallback, this);
        sub_odom_ = nh.subscribe("/rexrov2/pose_gt", 1, &SonarOccupancyMapper::odomCallback, this);
        pub_map_ = nh.advertise<nav_msgs::OccupancyGrid>("/projected_sonar_map", 1);

        ROS_INFO("Sonar Mapper C++ Started. Grid Size: %dx%d", grid_width_, grid_height_);
    }

private:
    // --- 成员变量 ---
    tf2_ros::Buffer tf_buffer_;
    tf2_ros::TransformListener tf_listener_;
    ros::Subscriber sub_cloud_;
    ros::Subscriber sub_odom_;
    ros::Publisher pub_map_;

    std::string map_frame_;
    std::string robot_base_frame_;
    
    float map_res_;
    float map_width_m_, map_height_m_;
    float map_origin_x_, map_origin_y_;
    int grid_width_, grid_height_, grid_size_;
    
    float sonar_max_range_;
    float sonar_fov_rad_;
    float hit_increment_;
    float max_occupancy_;
    float base_decay_rate_;
    float velocity_decay_factor_;
    int dilation_radius_;
    float max_tilt_deg_;

    std::vector<float> occupancy_grid_;
    float current_speed_;
    ros::Time last_update_time_;

    // --- 回调函数 ---
    void odomCallback(const nav_msgs::Odometry::ConstPtr& msg) {
        float vx = msg->twist.twist.linear.x;
        float vy = msg->twist.twist.linear.y;
        current_speed_ = std::sqrt(vx*vx + vy*vy);
    }

    void cloudCallback(const sensor_msgs::PointCloud2::ConstPtr& cloud_msg) {
        ros::Time current_time = ros::Time::now();
        double dt = (current_time - last_update_time_).toSec();
        if (dt < 0) dt = 0.0;
        last_update_time_ = current_time;

        // 1. 严格姿态检查 (Base Link Tilt Check)
        try {
            geometry_msgs::TransformStamped base_trans = tf_buffer_.lookupTransform(
                map_frame_, robot_base_frame_, cloud_msg->header.stamp, ros::Duration(0.1));
            
            tf2::Quaternion q(
                base_trans.transform.rotation.x,
                base_trans.transform.rotation.y,
                base_trans.transform.rotation.z,
                base_trans.transform.rotation.w);
            
            double roll, pitch, yaw;
            tf2::Matrix3x3(q).getRPY(roll, pitch, yaw);

            double max_tilt_rad = max_tilt_deg_ * M_PI / 180.0;
            if (std::abs(roll) > max_tilt_rad || std::abs(pitch) > max_tilt_rad) {
                publishMap(cloud_msg->header.stamp); // 仅发布旧地图
                return;
            }
        } catch (tf2::TransformException &ex) {
            return;
        }

        // 2. 获取传感器位姿
        geometry_msgs::TransformStamped sensor_trans_msg;
        try {
            sensor_trans_msg = tf_buffer_.lookupTransform(
                map_frame_, cloud_msg->header.frame_id, cloud_msg->header.stamp, ros::Duration(0.1));
        } catch (tf2::TransformException &ex) {
            return;
        }

        // 提取传感器位置和 Yaw (用于 FOV 计算)
        float sensor_x = sensor_trans_msg.transform.translation.x;
        float sensor_y = sensor_trans_msg.transform.translation.y;
        
        tf2::Quaternion qs(
            sensor_trans_msg.transform.rotation.x,
            sensor_trans_msg.transform.rotation.y,
            sensor_trans_msg.transform.rotation.z,
            sensor_trans_msg.transform.rotation.w);
        double s_roll, s_pitch, s_yaw;
        tf2::Matrix3x3(qs).getRPY(s_roll, s_pitch, s_yaw);

        // 3. 处理点云 (Hit Update)
        // 将 ROS 消息转换为 PCL 格式
        pcl::PointCloud<pcl::PointXYZ> pcl_cloud;
        pcl::fromROSMsg(*cloud_msg, pcl_cloud);

        // 转换点云到 Map Frame
        pcl::PointCloud<pcl::PointXYZ> cloud_world;
        Eigen::Isometry3d transform_eigen = tf2::transformToEigen(sensor_trans_msg);
        // pcl_ros::transformPointCloud 使用 Eigen::Matrix4f，需要转换
        pcl::transformPointCloud(pcl_cloud, cloud_world, transform_eigen.matrix().cast<float>());

        // 创建当前帧的 Hit Mask (避免同一帧对同一栅格重复计数)
        // 使用 vector<bool> 的特化版本或 vector<uint8_t>。uint8_t 更快避免位操作开销。
        std::vector<uint8_t> current_hits(grid_size_, 0);
        bool has_hits = false;

        for (const auto& pt : cloud_world.points) {
            if (std::isnan(pt.x) || std::isnan(pt.y)) continue;

            int cx = static_cast<int>((pt.x - map_origin_x_) / map_res_);
            int cy = static_cast<int>((pt.y - map_origin_y_) / map_res_);

            if (isValid(cx, cy)) {
                current_hits[cy * grid_width_ + cx] = 1;
                has_hits = true;
            }
        }

        if (has_hits) {
            applyHitsWithDilation(current_hits);
        }

        // 4. 应用 FOV 衰减 (Decay Update)
        applyFOVDecay(dt, sensor_x, sensor_y, static_cast<float>(s_yaw));

        // 5. 发布地图
        publishMap(cloud_msg->header.stamp);
    }

    // --- 核心算法实现 ---

    void applyHitsWithDilation(const std::vector<uint8_t>& hits) {
        // 如果没有膨胀，直接更新
        if (dilation_radius_ <= 0) {
            for (int i = 0; i < grid_size_; ++i) {
                if (hits[i]) {
                    occupancy_grid_[i] = std::min(occupancy_grid_[i] + hit_increment_, max_occupancy_);
                }
            }
            return;
        }

        // 带膨胀的更新
        // 为了性能，我们不创建完整的 dilated mask，而是遍历 hits，直接更新周围的 grid
        // 注意：这可能导致重叠区域被多次加分？原 Python 代码逻辑是先 dilate mask (bool OR) 再加分
        // 所以我们需要第二张 mask 或 set 来记录这一帧哪里被“膨胀后”击中了。
        
        std::vector<uint8_t> dilated_hits(grid_size_, 0);

        // 寻找所有 hit 点并标记膨胀区域
        // 遍历整个 hits 数组可能较慢，但比 Python 快。
        // 优化：我们可以在上面的点云循环中记录 hit 的索引，这里只遍历索引。
        // 但为了保持结构简单且 C++ 足够快，这里直接遍历 hits。
        
        for (int y = 0; y < grid_height_; ++y) {
            for (int x = 0; x < grid_width_; ++x) {
                if (hits[y * grid_width_ + x]) {
                    // 膨胀
                    for (int dy = -dilation_radius_; dy <= dilation_radius_; ++dy) {
                        for (int dx = -dilation_radius_; dx <= dilation_radius_; ++dx) {
                            int nx = x + dx;
                            int ny = y + dy;
                            if (isValid(nx, ny)) {
                                dilated_hits[ny * grid_width_ + nx] = 1;
                            }
                        }
                    }
                }
            }
        }

        // 应用增量
        for (int i = 0; i < grid_size_; ++i) {
            if (dilated_hits[i]) {
                occupancy_grid_[i] = std::min(occupancy_grid_[i] + hit_increment_, max_occupancy_);
            }
        }
    }

    void applyFOVDecay(double dt, float sensor_x, float sensor_y, float sensor_yaw) {
        if (dt <= 0) return;

        int range_indices = static_cast<int>(sonar_max_range_ / map_res_) + 1;
        int cx = static_cast<int>((sensor_x - map_origin_x_) / map_res_);
        int cy = static_cast<int>((sensor_y - map_origin_y_) / map_res_);

        int min_x = std::max(0, cx - range_indices);
        int max_x = std::min(grid_width_, cx + range_indices);
        int min_y = std::max(0, cy - range_indices);
        int max_y = std::min(grid_height_, cy + range_indices);

        if (min_x >= max_x || min_y >= max_y) return;

        float decay_amount = (base_decay_rate_ + current_speed_ * velocity_decay_factor_) * static_cast<float>(dt);
        float decay_max_range_sq = std::pow(sonar_max_range_ + 1.0f, 2);
        float half_fov = (sonar_fov_rad_ + (5.0f * M_PI / 180.0f)) / 2.0f;

        for (int y = min_y; y < max_y; ++y) {
            // 预计算 y 轴的世界坐标部分
            float world_y = y * map_res_ + map_origin_y_;
            float dy = world_y - sensor_y;

            for (int x = min_x; x < max_x; ++x) {
                float world_x = x * map_res_ + map_origin_x_;
                float dx = world_x - sensor_x;

                float dist_sq = dx * dx + dy * dy;

                if (dist_sq < decay_max_range_sq) {
                    float angle = std::atan2(dy, dx);
                    float delta_angle = angle - sensor_yaw;
                    
                    // 归一化角度到 -PI ~ PI
                    while (delta_angle <= -M_PI) delta_angle += 2 * M_PI;
                    while (delta_angle > M_PI) delta_angle -= 2 * M_PI;

                    if (std::abs(delta_angle) < half_fov) {
                        int idx = y * grid_width_ + x;
                        occupancy_grid_[idx] = std::max(0.0f, occupancy_grid_[idx] - decay_amount);
                    }
                }
            }
        }
    }

    void publishMap(ros::Time stamp) {
        nav_msgs::OccupancyGrid grid_msg;
        grid_msg.header.stamp = stamp;
        grid_msg.header.frame_id = map_frame_;
        grid_msg.info.resolution = map_res_;
        grid_msg.info.width = grid_width_;
        grid_msg.info.height = grid_height_;
        grid_msg.info.origin.position.x = map_origin_x_;
        grid_msg.info.origin.position.y = map_origin_y_;
        grid_msg.info.origin.position.z = -76.0;
        grid_msg.info.origin.orientation.w = 1.0;

        grid_msg.data.resize(grid_size_);
        
        // 转换 float -> int8
        for (int i = 0; i < grid_size_; ++i) {
            grid_msg.data[i] = static_cast<int8_t>(occupancy_grid_[i]);
        }

        pub_map_.publish(grid_msg);
    }

    inline bool isValid(int x, int y) {
        return (x >= 0 && x < grid_width_ && y >= 0 && y < grid_height_);
    }
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "sonar_occupancy_mapper");
    SonarOccupancyMapper node;
    ros::spin();
    return 0;
}